Develop > Interoperability > Introduction to XCM (#147)
polkadot-developers/polkadot-docs: develop/interoperability/intro-to-xcm.md

---
title: Introduction to XCM

> description: Unlock seamless blockchain interoperability with XCM—Polkadot's cross-consensus messaging format for asset transfers and cross-chain interactions.

Reviewer: ```suggestion
description: Unlock blockchain interoperability with XCM — Polkadot's Cross-Consensus Messaging format for cross-chain interactions.
```

I wouldn't say it's seamless 😂

---

# Introduction to XCM

## Introduction

Polkadot’s unique value lies in its ability to enable interoperability between parachains and other blockchain systems. At the core of this capability is XCM (Cross-Consensus Messaging)—a flexible messaging format that facilitates communication and collaboration between independent consensus systems.

> With XCM, blockchains can exchange assets, data, and instructions seamlessly, fostering a more interconnected ecosystem. Although it was developed specifically for Polkadot, XCM is a universal format, adaptable to any blockchain environment. This guide provides an overview of XCM’s core principles, design, and functionality, alongside practical examples of its implementation.

Reviewer: ```suggestion
With XCM, one chain can send intents to another one, fostering a more interconnected ecosystem. Although it was developed specifically for Polkadot, XCM is a universal format, usable in any blockchain environment. This guide provides an overview of XCM’s core principles, design, and functionality, alongside practical examples of its implementation.
```

## Messaging Format

> XCM is not a protocol but a standardized [messaging format](https://github.com/polkadot-fellows/xcm-format){target=\_blank}. It defines the structure and behavior of messages but does not handle their delivery. This separation allows developers to focus on crafting precise instructions for target systems without worrying about transmission mechanics.

Reviewer: ```suggestion
XCM is not a protocol but a standardized [messaging format](https://github.com/polkadot-fellows/xcm-format){target=\_blank}. It defines the structure and behavior of messages but does not handle their delivery. This separation allows developers to focus on crafting instructions for target systems without worrying about transmission mechanics.
```

> XCM messages are action-oriented, specifying how the receiving blockchain should alter its state. Unlike transactions, they do not directly perform these changes. By leveraging asynchronous composability, XCM enables efficient, out-of-sequence execution, akin to how RESTful services structure HTTP requests.

Reviewer: What do you mean by this? Out-of-sequence?
Also XCMs just have intents, the underlying changes are decided by the host environment (the chains).

Reviewer 2: agree, will paraphrase this paragraph just a bit to align it with francisco's comment :)

---
title: Introduction to XCM

> description: A basic introduction to the world of XCM. Covering a range of topics that will help you get started with learning XCM and its usecases.
> ---
> # Introduction to XCM

Reviewer 2: ```suggestion
description: A basic introduction to the world of XCM, covering a range of topics that will help you get started with learning XCM and its use cases.
---

# Introduction to XCM
```

## What is 'XCM' - Cross-Consensus Messaging?

> XCM (Cross Consensus Messaging) is a language that enables communication between chains in the Polkadot ecosystem. XCM allows blockchains to seamlessly send messages to each other, enabling the transfer of assets, data, and other information across different chains and smart contracts. XCM was devleoped within the Polkadot ecosystem, but is designed to be general enough to provide a common format for cross-consensus communication that can be used anywhere. This document will provide you with an introduction to XCM, covering the basics to help you futher understand the language.

Reviewer 2: I think we can add a link to the XCM repo here:

```suggestion
[XCM (Cross Consensus Messaging)](https://github.com/polkadot-fellows/xcm-format){target=\_blank} is a language that enables communication between chains in the Polkadot ecosystem. XCM allows blockchains to seamlessly send messages to each other, enabling the transfer of assets, data, and other information across different chains and smart contracts. XCM was developed within the Polkadot ecosystem but is designed to be general enough to provide a common format for cross-consensus communication that can be used anywhere. This document will provide you with an introduction to XCM, covering the basics to help you further understand the language.
```

One of Polkadot's main functionalities is interoperability amongst parachains and any other participating consensus-driven systems. XCM is the language through which complex, cross-consensus interactions can occur. Two blockchains can "speak" XCM to seamlessly interact with each other using a standard messaging format.

> ## A format, not a protocol
> It's important to remember that XCM is a messaging format, and not a protocol. It cannot be used to 'send messages' but instead express what should be done by the receiver. XCM does not define how messages are delivered but rather define how they should look, act, and contain relative instructions to the on-chain actions the message intends to perform.

Reviewer 2: ```suggestion
## A format, not a protocol

It's important to remember that XCM is a [messaging format](https://github.com/polkadot-fellows/xcm-format){target=\_blank}, not a protocol. It cannot be used to 'send messages' but instead express what should be done by the receiver. XCM does not define how messages are delivered but rather how they should look, act, and contain relative instructions to the on-chain actions the message intends to perform.
```

XCM messagges, by themselves, are not considered transactions. XCM just describes how to change the state of the target network, but the message by itself doesn't perform the state change. 

This partly ties to what is called asynchronous composability, which allows XCM messages to bypass the concept of time-constrained mechanisms, like on-chain scheduling and execution over time in the correct order in which it was intended.

XCM works similarly to how RESTful services use REST as an architectural style of development, where HTTP requests contain specific parameters to perform an action.

> ## The Four Principles of XCM
> XCM has four high-level core design principles which it stands to follow:
> 
> * **Asynchronous**
> 
>     XCM messages in no way assume that the sender will be blocking on its completion.
> 
> * **Absolute**
> 
>     XCM messages are guaranteed to be delivered and interpreted accurately, in order and in a timely fashion. Once a message is sent, one can be sure it will be processed as it was intended to be.
> 
> * **Asymmetric**
> 
>     XCM messages, by default, do not have results that let the sender know that the message was received - they follow the 'fire and forget' paradigm. Any results must be separately communicated to the sender with an additional message back to the origin.
> 
> * **Agnostic**
> 
>     XCM makes no assumptions about the nature of the consensus systems between which the messages are being passed. XCM as a message format should be usable in any system that derives finality through consensus.

Reviewer 2: ```suggestion
## The Four Principles of XCM

XCM has four high-level core design principles which it stands to follow:

- **Asynchronous** - XCM messages do not assume the sender will be blocking on its completion
- **Absolute** - XCM messages are guaranteed to be delivered and interpreted accurately, in an order, and timely fashion. Once a message is sent, one can be sure it will be processed as intended
- **Asymmetric** - XCM messages, by default, do not provide results that inform the sender that the message was received; they follow the 'fire and forget' paradigm. Any results must be communicated separately to the sender with an additional message back to the origin
- **Agnostic** - XCM makes no assumptions about the nature of the consensus systems between which the messages are passed. XCM, as a message format, should be usable in any system that derives finality through consensus
```

These four crucial design decisions allow for XCM messages to be a reliable yet convenient way to properly convey the intentions from one consensus system to another without any compatibility issues.

## The XCM Tech Stack

> 

Reviewer 2: If possible, I'd like a small paragraph that explains all layers, at a high level of course 🙏

Author: Makes sense, was also thinking this today, will work on it shortly!

<figure markdown>
  <img src="../../../images/develop/parachain-devs/interoperability/xcm-tech-stack/cross-consensus-tech-stack-e9c1b2ab8b6f6f3f9a78b3a412af0698.png" width="800">
  <figcaption>The XCM Tech Stack</figcaption>
</figure>

## Core Functionalities of XCM
Some of the key features and additions to cross-consensus messaging include

* **Programmability**

    The ability to have expectations for messages, which allow for more comprehensive use cases, safe dispatches for version checking, branching, and NFT/Asset support.

* **Functional Multichain Decomposition**

    The ability to define mechanisms to cross-reference and perform actions on other chains on behalf of the origin chain (remote locking), context/id for these messages, and asset namespacing.

* **Bridging**

    Introduces the concept of a universal location, which allows for a base reference for global consensus systems for multi-hop setups. This location is above the parent relay chain or other consensus systems like Ethereum or Bitcoin.

> A core part of the vision that XCM provides is improving communication between the chains to make system parachains a reality. For example, the Polkadot relay chain handles more than just parachain management and shared security - it handles user balances/assets, governance, and staking. Ideally, the relay chain should be for what it's intended to be - a place for shared security. System parachains can alleviate these core responsibilities from the relay chain but only by using a standard format like XCM.

Reviewer 2: I'm unsure if it makes sense to say that the Polkadot relay chain handles "assets", for balances it's okay but saying asset might be misleading with asset hub (or plaza in the future)

Author: Happy to change it! What would we suggest changing it too?

Reviewer 2: Thanks! I'd remove it, like: "...it handles user balances, governance, and staking"

WDYT @RemyLeBerre ?

Reviewer 2: oh seems to be fixed, solving this conversation

> This is where system parachains come in, where each of these core responsibilities can be delegated to a system parachain respectively.

Reviewer 2: I'd remove this paragraph, since the plaza introduction would make it so that not 'each' responsibility will be delegated to a different parachain, just to be consistent..

> ## XCM Example
> Below shows a simple example of an XCM message. In the example, the native token is withdrawn from the account of Alice and deposited into the account of Bob. The message simulates a transfer between the two accoubts in the same consensus system (ParaA).  You can find the complete example in the [repo](https://github.com/paritytech/xcm-docs/tree/main/examples){target=_blank}.

Reviewer 2: This is a nitpick comment, but maybe it would be nice to create a note below this rust snippet with this sentence: "You can find the complete example in..."

```suggestion
## XCM Example

A simple example of an XCM message is shown below. In the example, the native token is withdrawn from Alice's account and deposited into Bob's account. The message simulates a transfer between the two accounts in the same consensus system (`ParaA`). You can find the complete example in the [XCM Docs](https://github.com/paritytech/xcm-docs/tree/main/examples){target=\_blank} repo.
```

Reviewer 2: Also, it would make sense to add a direct reference to the Withdraw instruction - https://github.com/polkadot-fellows/xcm-format?tab=readme-ov-file#withdrawasset

Author: Added!

```rust
--8<-- 'code/develop/parachain-devs/interoperability/XCM/XCM-first-look.rs'

> ```
> The message consists of three instructions

Reviewer 2: ```suggestion
```

The message consists of three instructions
```

> The message consists of three instructions

Reviewer 2: ```suggestion

The message consists of three instructions
```

* **WithdrawAsset**
```rust
--8<-- 'code/develop/parachain-devs/interoperability/XCM/XCM-withdrawasset.rs'
```

>     The first instruction takes as an input the MultiAsset that should be withdrawn. The MultiAsset describes the native parachain token with the `Here` keyword. The `amount` parameter is the number of tokens that are transferred. The withdrawal account depends on the origin of the message. In this example the origin of the message is Alice. The `WithdrawAsset` instruction moves `amount` number of native tokens from Alice's account into the holding register.

Reviewer 2: Maybe it's worth adding a reference to multi-asset as well.

* **BuyExecution**
```rust
--8<-- 'code/develop/parachain-devs/interoperability/XCM/XCM-buyexecution.rs'
```

>     To execute XCM instructions, weight (some amount of resources) has to be bought. The amount of weight needed to execute an XCM depends on the number and type of instructions in the XCM. The `BuyExecution` instruction pays for the weight using the `fees`. The `fees` parameter describes the asset in the holding register that should be used for paying for the weight. The `weight_limit` parameter defines the maximum amount of fees that can be used for buying weight. There are special occasions where it is not necessary to buy weight. 

Reviewer 2: ```suggestion
  
      To execute XCM instructions, weight (some resources) must be bought. The weight needed to execute an XCM depends on the number and type of instructions in the XCM. The `BuyExecution` instruction pays the weight using the `fees`. The `fees` parameter describes the asset in the holding register that should be used to pay for the weight. The `weight_limit` parameter defines the maximum fees that can be used to buy weight. There are special occasions where it is not necessary to buy weight. 
```

* **DepositAsset**
```rust
--8<-- 'code/develop/parachain-devs/interoperability/XCM/XCM-depositasset.rs'
```

>     The `DepositAsset` instruction is used to deposit funds from the holding register into the account of the beneficiary. You dont actually know how much is remaining in the holding register after the `BuyExecution` instruction, but that does not matter since you can specify a wildcard for the asset(s) which should be deposited.

Reviewer 2: add reference to:
- DepositAsset - https://github.com/polkadot-fellows/xcm-format?tab=readme-ov-file#depositasset
- BuyExecution - https://github.com/polkadot-fellows/xcm-format?tab=readme-ov-file#buyexecution

Author: Done & done, ty!

    
    In this case, the wildcard is `All`, meaning that all assets in the holding register at that point in the execution should be deposited. The beneficiary in this case is the account of Bob in the current consensus system. 
    
    When the three instructions are combined, you withdraw `amount` native tokens from the account of Alice, pay for the execution of these instructions, and deposit the remaining tokens in the account of Bob.

> ## Overview
> XCM enables different consensus systems to communicate with each other. Common cross-consensus use-cases include:
> 
> * Sending tokens between blockchains
> * Locking assets on one blockchain in order to gain some benefit on a smart contract on another blockchain
> * Calling specific functions on another blockchain

Reviewer 2: ```suggestion
## Overview

XCM enables different consensus systems to communicate with each other. Common cross-consensus use-cases include:

- Sending tokens between blockchains
- Locking assets on one blockchain to gain some benefit on a smart contract on another blockchain
- Calling specific functions on another blockchain
```

These are just a few basic examples; once you can communicate with other consensus systems, you can create applications that can leverage multiple blockchains' capabilities. The potential it provides is especially evident in an ecosystem of highly specialized blockchains like Polkadot.

Decentralized distributed systems are very complex, so it's easy to make errors when building interactions between them. XCM is meant to be used by developers to package these interactions into their runtime logic before exposing that functionality to end users.

> XCM is constantly evolving; the format is expected to change over time. To keep up with the development of the format, or to propose changes, go to the [XCM format repository](https://github.com/polkadot-fellows/xcm-format).

Reviewer 2: I'd add a small sentence that explains XCM versions (what's the current version used or something like that), since you are touching on the evolution of XCM
